extension/asapclientauthextension/extension.go

```diff
@@ -0,0 +1,96 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package asapclientauthextension
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	asap "bitbucket.org/atlassian/go-asap"
+	"github.com/SermoDigital/jose/crypto"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/config/configauth"
+	"google.golang.org/grpc/credentials"
+)
+
+// AsapClientAuthenticator implements ClientAuthenticator
+type AsapClientAuthenticator struct {
+	provisioner asap.Provisioner
+	privateKey  interface{}
+}
+
+func (a AsapClientAuthenticator) RoundTripper(base http.RoundTripper) (http.RoundTripper, error) {
+	return asap.NewTransportDecorator(a.provisioner, a.privateKey)(base), nil
+}
+
+func (a AsapClientAuthenticator) PerRPCCredentials() (credentials.PerRPCCredentials, error) {
+	token, err := a.provisioner.Provision()
+	if err != nil {
+		return nil, err
+	}
+	headerValue, err := token.Serialize(a.privateKey)
+	if err != nil {
+		return nil, err
+	}
+	header := map[string]string{
+		"authorization": fmt.Sprintf("Bearer %s", string(headerValue)),
+	}
+	return &PerRPCAuth{
+		metadata: header,
+	}, nil
+}
+
+func (a AsapClientAuthenticator) Start(_ context.Context, _ component.Host) error {
+	return nil
+}
+
+func (a AsapClientAuthenticator) Shutdown(_ context.Context) error {
+	return nil
+}
+
+var _ configauth.ClientAuthenticator = (*AsapClientAuthenticator)(nil)
+
+func createAsapClientAuthenticator(cfg *Config) (AsapClientAuthenticator, error) {
+	var a AsapClientAuthenticator
+	pk, err := asap.NewPrivateKey([]byte(cfg.PrivateKey))
+	if err != nil {
+		return a, err
+	}
+	a = AsapClientAuthenticator{
+		provisioner: asap.NewCachingProvisioner(asap.NewProvisioner(
+			cfg.KeyId, time.Duration(cfg.Ttl)*time.Second, cfg.Issuer, cfg.Audience, crypto.SigningMethodRS256)),
+		privateKey: pk,
+	}
+	return a, nil
+}
+
+var _ credentials.PerRPCCredentials = (*PerRPCAuth)(nil)
+
+// PerRPCAuth is a gRPC credentials.PerRPCCredentials implementation that returns an 'authorization' header.
+type PerRPCAuth struct {
```

Comment:
PerRPCAuth is essentially copied from the bearer auth extension. Also it implements gRPC's `credentials.PerRPCCredentials`, is it possible to do with a map? and how does it reduce allocations?

Reply:
Since go is pass by copy, wrapping a map around a struct does have some overhead, however, casting to and from type that extends a map would incur a smaller memory footprint.